mvcc/kvstore.go

```diff
@@ -245,9 +245,10 @@ func (s *store) Compact(rev int64) (<-chan struct{}, error) {
 	// ensure that desired compaction is persisted
 	s.b.ForceCommit()
 
-	keep := s.kvindex.Compact(rev)
 	ch := make(chan struct{})
 	var j = func(ctx context.Context) {
+		keep := s.kvindex.Compact(rev)
```

Comment:
why this needs to be changed?

Reply:
The defer of `s.mu.Unlock()` seems to mean that throughput still stalls during index compaction, although on audit of the code it isn't clear how this is the case. We can look to see if we can create a test to validate that.

Reply:
i mean why do we move the compact func into the schedule j unit?

Reply:
It seemed cleaner to put it there rather than remove the deferred unlock in place of explicitly unlocking in the error cases. This could be mistaken.

Reply:
[Transactions](https://github.com/jcalvert/etcd/blob/48554b490a3fe57c6ae3f221f968f0835fb758ad/mvcc/kvstore_txn.go#L31) do indeed touch this lock so by offloading it to the scheduled action it means that transactions won't time out waiting for the lock. If you'd prefer us to change it so that the code explicitly unlocks we can do that.

Reply:
i do not think this is safe since this func expects to be returned after the index is compacted (the old revisions are not reachable). we still need to keep the index compaction synchronously